contracts/p1/RevenueTrader.sol

```diff
@@ -89,14 +119,121 @@ contract RevenueTraderP1 is TradingP1, IRevenueTrader {
         );
 
         if (launch) {
-            tryTrade(req);
+            openTrade(req);
+        }
+    }
+
+    /// Executes an atomic swap for revenue via a dutch auction
+    /// @dev Caller must have granted tokenIn allowances for required tokenIn bal
+    /// @param tokenIn The ERC20 token provided by the caller
+    /// @param tokenOut The ERC20 token being purchased by the caller
+    /// @param amountOut {qTokenOut} The exact quantity of tokenOut being purchased
+    /// @return s The exact Swap performed
+    /// @custom:interaction RCEI
+    function swap(
+        IERC20 tokenIn,
+        IERC20 tokenOut,
+        uint256 amountOut
+    ) external notTradingPausedOrFrozen returns (Swap memory s) {
+        // == Refresh ==
+        assetRegistry.refresh();
+        furnace.melt();
+
+        require(address(trades[tokenOut]) == address(0), "nonatomic trade ongoing");
+        require(tokenIn == tokenToBuy, "will only buy tokenToBuy");
+        require(tokenOut != tokenToBuy, "will not sell tokenToBuy");
+
+        // executeSwap if storage auction already exists
+        DutchAuction storage auction = dutchAuctions[tokenOut][tradeEnd];
+        if (
+            tradeEnd > block.timestamp &&
+            (address(auction.sell) != address(0) || address(auction.buy) != address(0))
+        ) {
+            return executeSwap(auction, amountOut);
+        }
+
+        // === Checks/Effects ===
+
+        require(
+            tradeEnd + dutchAuctionLength > block.timestamp &&
+                block.timestamp + dutchAuctionLength > tradeEnd,
```

Comment:
i think we want >= here

Reply:
edge case: using `>` makes it so that multiple auctions cannot be swapped in the same block. the first would bump `tradeEnd` (below), and the next would fail here